pygeos/utils.py

```diff
@@ -1,14 +1,27 @@
 from contextlib import contextmanager
+from itertools import chain
+from functools import wraps
 import numpy as np
 
-@contextmanager
-def not_writable(arr):
-    if isinstance(arr, np.ndarray):
-        old = arr.flags.writeable
-        arr.flags.writeable = False
+
+def multithreading_enabled(func):
+    """Prepare multithreading by setting the writable flags of object type
+    ndarrays to False.
+
+    NB: multithreading also requires the GIL to be released, which is done in
+    the C extension (ufuncs.c)."""
+    @wraps(func)
+    def wrapped(*args, **kwargs):
+        array_args = [
+            arg for arg in chain(args, kwargs.values())
+            if isinstance(arg, np.ndarray) and arg.dtype == object
+        ]
```

Comment:
Alternatively, we could also specify the keywords which needs to be treated this way in the decorator. Like:

```
@multithreading_enabled(['a', 'b'])
def distance(a, b, **kwargs):
...

@multithreading_enabled(['geometry'])
def length(geometry, **kwargs):
...
``` 

That's a bit more work to write, but on the other hand is more explicit and avoids inference based on the dtype

Reply:
Although, since you can also pass a single scalar, you would still need to check whether it is an array anyway .. 
And since the arguments can both be passed as arg or kwarg, getting the values out of that based on the position or name, will also get a bit complicated ..

(python only supports creating position only arguments starting from 3.8, otherwise we could have used it here)

Reply:
Yes, followed the same train of thought. So leave it like this? What about the name? It feels kind of misleading now.

May be `lock_object_ndarrays`?

Reply:
Yeah, leaving it like this is fine for me.

About the name, I like the fact that it is clear what the *reason* is that we use this (because we "allow threads", as the pythn C api says), but it's true it doesn't really reflect *what it does*, so can go both ways